pkg/daemon/vm_controller.go

```diff
@@ -75,56 +75,66 @@ func (r *VMReconciler) reconcile(ctx context.Context, vm *virtv1alpha1.VirtualMa
 				if err := r.getCloudHypervisorClient(vm).VmShutdown(ctx); err != nil {
 					r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedPowerOff", "Failed to powered off VM")
-				} else {
-					r.Recorder.Eventf(vm, corev1.EventTypeNormal, "PoweredOff", "Powered off VM")
-				}
-			case virtv1alpha1.VirtualMachineShutdown:
-				if err := r.getCloudHypervisorClient(vm).VmPowerButton(ctx); err != nil {
-					r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedShutdown", "Failed to shutdown VM")
-				} else {
-					r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Shutdown", "Shutdown VM")
-				}
-			case virtv1alpha1.VirtualMachineReset:
-				if err := r.getCloudHypervisorClient(vm).VmReboot(ctx); err != nil {
-					r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedReset", "Failed to reset VM")
-				} else {
-					r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Reset", "Reset VM")
-				}
-			case virtv1alpha1.VirtualMachineReboot:
-				// TODO: reboot
-				if err := r.getCloudHypervisorClient(vm).VmReboot(ctx); err != nil {
-					r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedReboot", "Failed to reboot VM")
-				} else {
-					r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Rebooted", "Rebooted VM")
+					return fmt.Errorf("power off VM: %s", err)
 				}
-			case virtv1alpha1.VirtualMachinePause:
-				if err := r.getCloudHypervisorClient(vm).VmPause(ctx); err != nil {
-					r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedPause", "Failed to pause VM")
-				} else {
-					r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Paused", "Paused VM")
+			} else {
+				switch vm.Status.PowerAction {
+				case virtv1alpha1.VirtualMachinePowerOff:
+					if err := r.getCloudHypervisorClient(vm).VmShutdown(ctx); err != nil {
+						r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedPowerOff", "Failed to powered off VM")
+					} else {
+						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "PoweredOff", "Powered off VM")
+					}
+				case virtv1alpha1.VirtualMachineShutdown:
+					if err := r.getCloudHypervisorClient(vm).VmPowerButton(ctx); err != nil {
+						r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedShutdown", "Failed to shutdown VM")
+					} else {
+						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Shutdown", "Shutdown VM")
+					}
+				case virtv1alpha1.VirtualMachineReset:
+					if err := r.getCloudHypervisorClient(vm).VmReboot(ctx); err != nil {
+						r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedReset", "Failed to reset VM")
+					} else {
+						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Reset", "Reset VM")
+					}
+				case virtv1alpha1.VirtualMachineReboot:
+					// TODO: reboot
+					if err := r.getCloudHypervisorClient(vm).VmReboot(ctx); err != nil {
+						r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedReboot", "Failed to reboot VM")
+					} else {
+						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Rebooted", "Rebooted VM")
+					}
+				case virtv1alpha1.VirtualMachinePause:
```

Comment:
this logic is not reentrant, test with `ch-remote pause` twice, the latter one will report an error.

Reply:
Let's fix it in another patch?